src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala

```diff
@@ -253,7 +253,12 @@ abstract class BCodeHelpers extends BCodeIdiomatic {
       def fail(msg: String, pos: Position = sym.pos) = {
         reporter.warning(sym.pos,
           sym.name +
-          s" has a main method with parameter type Array[String], but ${sym.fullName('.')} will not be a runnable program.\n  Reason: $msg"
+          s""" has a main method with parameter type Array, but ${sym.fullName('.')} will not be a runnable program.
+          |A runnable program is an object with a method main that has the exact
+          |signature (Array[String])Unit, doesn't have a trait as a companion class,
+          |doesn't have a companion class with a method called main, has no type
+          |paremters, and doesn't inherit or override anything.
+          |This fails because: $msg""".stripMargin
```

Comment:
I would keep the actual reason in `msg` closer to the top. I'm not sure the user is helped much with the full list of requirements for a main method, especially since we're already telling them what's wrong in this particular case.

Reply:
On further reflection, I like the way javac in som's snippet provides actionable info: "define the method as follows:..." (rather than listing the issues that make scalac's job hard, tell the user how to fix their problem :-))

Reply:
I meant to list the actual things wrong, instead of all possible requirements. I meant to PR but had a busy weekend. I realize it may be out of interest if not out of scope.

Reply:
If not 2.12.7, it'll be in 2.12.8 (more likely)